File: python/src/problem/leetcode/easy/leetcode_100.py
# coding=utf-8
from typing import List

from src.data_structure.data_structure import TreeNode


class Solution:
    """
    判断二叉树是否相同
    """

    def is_same_tree_recursive(self, p: TreeNode, q: TreeNode) -> bool:
        """
        Time: O(n), Space: O(n)
        :param p:
        :param q:
        :return:
        """
        if not p and not q:
            return True
        if not p or not q:
            return False
        return p.val == q.val and self.is_same_tree_recursive(p.left, q.left)\
               and self.is_same_tree_recursive(p.right, q.right)

    def is_same_tree_iterative(self, p: TreeNode, q: TreeNode) -> bool:
        """
        Time: O(n), Space: O(n)
        :param p:
        :param q:
        :return:
        """
        if not p and not q:
            return True
        if not p or not q:
            return False

        stack = [p, q]
        while len(stack) > 0:
            left = stack.pop()
            right = stack.pop()
            if not left and not right:
                continue
            if not left or not right or left.val != right.val:
                return False
            stack.append(left.left)
            stack.append(right.left)
            stack.append(left.right)
            stack.append(right.right)
        return True

File: java/src/main/java/com/ywh/problem/leetcode/easy/LeetCode100.java
package com.ywh.problem.leetcode.easy;

import com.ywh.ds.tree.TreeNode;

import java.util.Stack;

/**
 * 相同的树
 * [树] [栈]
 *
 * 给定两个二叉树，编写一个函数来检验它们是否相同。
 * 如果两个树在结构上相同，并且节点具有相同的值，则认为它们是相同的。
 * 示例 1:
 *      输入:       1         1
 *                / \       / \
 *               2   3     2   3
 *
 *              [1,2,3],   [1,2,3]
 *      输出: true
 * 示例 2:
 *      输入:      1          1
 *                /           \
 *               2             2
 *
 *              [1,2],     [1,null,2]
 *      输出: false
 * 示例 3:
 *      输入:       1         1
 *                / \       / \
 *               2   1     1   2
 *
 *              [1,2,1],   [1,1,2]
 *      输出: false
 *
 * @author ywh
 * @since 2/14/2019
 */
public class LeetCode100 {

    /**
     * 递归解法
     * Time: O(n), Space: O(n)
     *
     * @param p
     * @param q
     * @return
     */
    public boolean isSameTreeRecursive(TreeNode p, TreeNode q) {
        if (p == null && q == null) {
            return true;
        }
        if (p == null || q == null) {
            return false;
        }
        return p.val == q.val && isSameTreeIterative(p.left, q.left) && isSameTreeIterative(p.right, q.right);
    }

    /**
     * 迭代解法
     * Time: O(n), Space: O(n)
     *
     * @param p
     * @param q
     * @return
     */
    public boolean isSameTreeIterative(TreeNode p, TreeNode q) {
        Stack<TreeNode> stack = new Stack<>();
        stack.push(p);
        stack.push(q);
        while (!stack.isEmpty()) {
            TreeNode r = stack.pop(), l = stack.pop();
            if (l == null && r == null) {
                continue;
            }
            if (l == null || r == null || l.val != r.val) {
                return false;
            }
            stack.push(l.left);
            stack.push(r.left);
            stack.push(l.right);
            stack.push(r.right);
        }
        return true;
    }
}
